File: 0145.BinaryTreePostorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # Iteratively with stack
        ans = []
        stack = []
        node = root
        while stack or node:
            if node:
                stack.append(node)
                ans.append(node.val)
                node = node.right
            else:
                node = stack.pop()
                node = node.left
        return ans[::-1]
        '''
        # Recursive
        if not root:
            return []
        #return self.postorderTraversal(root.left) + self.postorderTraversal(root.right) + [root.val]
        ans = []
        ans += self.postorderTraversal(root.left)
        ans += self.postorderTraversal(root.right)
        ans.append(root.val)
        return ans
        '''

File: 0000-0999/0145.BinaryTreePostorderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func postorderTraversal(root *TreeNode) []int {
    ans := []int{}
    stack := []*TreeNode{}
    for len(stack) > 0 || root != nil {
        if root != nil {
            ans = append(ans, root.Val)
            stack = append(stack, root)
            root = root.Right
        } else {
            root = stack[len(stack)-1].Left
            stack = stack[:len(stack)-1]
        }
    }
    for i, j := 0, len(ans)-1; i < j; i, j = i+1, j-1 {
        ans[i], ans[j] = ans[j], ans[i]
    }
    return ans
}
